### PR TITLE
sandbox code optimization

### PR DESCRIPTION
**Pull request by dierbei**
There still seems to be a problem, see details: https://github.com/rust-lang/rust-clippy/issues/11247

In the future, I will rebase after the bug fix is completed to keep the commit record clear.

### CODE REVIEW

**Comment by dierbei** [comment]
Now that features have been removed, the project can be built with the following command:
- `cargo build --bin qemu`
- `cargo build --bin stratovirt`
- `cargo build --bin cloud_hypervisor`

**Comment by dierbei** [comment]
Seems like it could be further optimized.

 I will try to move `qumu`, `stratovirt ` and `cloud_hypervisor ` to separate folders under bin.

The advantage of this is that you can use the command directly: `cargo build`

**Comment by dierbei** [comment]
rust-analyzer in vscode prompts an error message:

```shell
[ERROR rust_analyzer::main_loop] FetchBuildDataError:
error[E0635]: unknown feature `proc_macro_span_shrink`
  --> /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/proc-macro2-1.0.51/src/lib.rs:92:30
   |
92 |     feature(proc_macro_span, proc_macro_span_shrink)
   |                              ^^^^^^^^^^^^^^^^^^^^^^


error: aborting due to previous error


For more information about this error, try `rustc --explain E0635`.

error: could not compile `proc-macro2` (lib) due to 2 previous errors

error[E0635]: unknown feature `proc_macro_span_shrink`
  --> /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/proc-macro2-1.0.56/src/lib.rs:92:30
   |
92 |     feature(proc_macro_span, proc_macro_span_shrink)
   |                              ^^^^^^^^^^^^^^^^^^^^^^


error: aborting due to previous error


For more information about this error, try `rustc --explain E0635`.

error: could not compile `proc-macro2` (lib) due to 2 previous errors
```

**Comment by Burning1020** [comment]
Will take a look.

File: .gitignore [hunk 1]
```diff
@@ -5,3 +5,4 @@ target
 **/src/api/*
 !**/src/api/mod.rs
 vmm/common/Cargo.lock
+!/vmm/sandbox/src/bin/
```
Comment (Burning1020) [inline .gitignore:8]: Is it helpful to replace Line2 `bin` with `/bin`

File: vmm/sandbox/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+pub const NAMESPACE_NET: &str = "network";
+pub const NAMESPACE_MNT: &str = "mount";
+pub const NAMESPACE_CGROUP: &str = "cgroup";
+
+pub const FS_SHARE_PATH: &str = "shared_fs";
+
+pub const CONFIG_STRATOVIRT_PATH: &str = "/var/lib/kuasar/config_stratovirt.toml";
+pub const CONFIG_CLH_PATH: &str = "/var/lib/kuasar/config_clh.toml";
+
+pub async fn init_cloud_hypervisor_sandboxer(
```
Comment (Burning1020) [inline vmm/sandbox/src/lib.rs:63]: why is this init function needed? maybe move it to cloud_hypervisor folder?

File: vmm/sandbox/Cargo.toml [hunk 3]
```diff
@@ -7,12 +7,6 @@ edition = "2021"
 [profile.release]
 panic = 'abort'
 
-[features]
-default = []
-qemu = ["qapi/qmp", "qapi/async-tokio-all", "unshare", "os_pipe"]
```
Comment (Burning1020) [inline vmm/sandbox/Cargo.toml:12]: Please try to add the mapping of binary name and its main.rs path. https://doc.rust-lang.org/cargo/reference/cargo-targets.html#configuring-a-target

**Comment by Burning1020** [comment]
One more thing is to check Makefile targets.

File: .gitignore (hunk 1 above)
Comment (dierbei) [reply .gitignore]: Thanks for the advice. I made the changes.

`/bin` ignores all `bin` directories in the project directory.

File: vmm/sandbox/src/lib.rs (hunk 2 above)
Comment (dierbei) [reply vmm/sandbox/src/lib.rs]: I tried moving the initialization functions for `qemu`,`cloud_hypervisor`, and `stratovirt` in `lib.rs` to their own folders.

File: vmm/sandbox/Cargo.toml (hunk 3 above)
Comment (dierbei) [reply vmm/sandbox/Cargo.toml:12]: I added configuration for qemu, cloud_hypervisor, stratovirt.

**Comment by dierbei** [comment]
> One more thing is to check Makefile targets.

I added the `--bin` argument to the `bin/vmm-sandboxer` fragment in the Makefile.

File: vmm/sandbox/src/bin/cloud_hypervisor/main.rs [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    // Initialize sandboxer
+    let sandboxer = init_cloud_hypervisor_sandboxer().await?;
+
+    // If 'log_level' field isn't set in the config file, keep the log level from the default env
+    // Otherwise, set the log level configured in the config file
+    if !sandboxer.log_level().is_empty() {
+        let log_level = log::LevelFilter::from_str(sandboxer.log_level())?;
+        builder.filter_level(log_level);
+    }
+    builder.init();
```
Comment (Burning1020) [inline vmm/sandbox/src/bin/cloud_hypervisor/main.rs:36]: Should we extra the common part of logger to a new function `init_logger()` and pass variable `sandboxer.log_level()`, something like this:
```rust
...
let sandboxer = init_cloud_hypervisor_sandboxer().await?;
init_logger(sandboxer.log_level());
...
```

**Review by Burning1020 — COMMENTED** [review]
/cc @abel-von @Vanient @flyflypeng

File: wasm/src/sandbox.rs [hunk 5]
```diff
@@ -94,7 +94,7 @@ impl Sandboxer for WasmSandboxer {
             .await
             .get(id)
             .ok_or_else(|| Error::NotFound(id.to_string()))?
-            .clone());
+            .clone())
```
Comment (abel-von) [inline wasm/src/sandbox.rs:97]: maybe "return" can also be removed

File: wasm/src/sandbox.rs [hunk 6]
```diff
@@ -214,7 +214,7 @@ impl Sandbox for WasmSandbox {
     async fn container(&self, id: &str) -> Result<&Self::Container> {
         return self.containers.get(id).ok_or(Error::NotFound(format!(
             "failed to find container by id {id}"
-        )));
+        )))
```
Comment (abel-von) [inline wasm/src/sandbox.rs:217]: it's better remove "return" too.

File: vmm/sandbox/src/bin/cloud_hypervisor/main.rs (hunk 4 above)
Comment (abel-von) [reply vmm/sandbox/src/bin/cloud_hypervisor/main.rs:36]: the three main.rs files are too many codes in common, maybe we can make some abstraction to remove duplicated codes.

**Comment by dierbei** [comment]
@Burning1020 @abel-von Thanks for the suggestion, I'll revise it soon.

**Review by Burning1020 — APPROVED** [review]
Good job. Thank you.

**Pull request merged**